### DISCUSSION TITLE
Drive state from jupyter notebook, with change reflected automatically in app

### DISCUSSION THREAD

**Discussion by cardinalgeo** [Kitware/trame#314] [Q&A] [answered]
Hi Trame team! 

I'd like to start up a trame application in a jupyter notebook, then alter the state using a notebook cell so that the change is reflected in the application. In other words, I'd like to programmatically interact with the app live. A use case in which this would be relevant is if I had a large list of items (mapped to cells or points) that I'd like to highlight in a scene. I could use the UI to select each item in a dropdown, which would be tedious, or I could feed a list of IDs for those items programmatically. 

I've figured out how to drive state from a cell, but I can't figure out how to have the application update to reflect this change without re-running the application's cell. I've created an MRE by modifying the trame "cheatsheet" example (from [here](https://kitware.github.io/trame/docs/cheatsheet.html)) to illustrate this dilemma. How would you suggest I proceed? 

To be pasted into a jupyter notebook: 

cell 1: 
```python
from trame.app import get_server, jupyter               # Entry point to trame
from trame.ui.vuetify import SinglePageLayout  # UI layout
from trame.widgets import vuetify, vtk         # UI widgets

server = get_server()                          # Create/retrieve default server
server.client_type = "vue2"                    # Choose between vue2 and vue3
state, ctrl = server.state, server.controller  # Extract server state and controller

# Reset resolution variable to 6
def reset_resolution():
    state.resolution = 6

# When resolution change, execute fn
@state.change("resolution")
def resolution_change(resolution, **kwargs):
    print(f"Slider updating resolution to {resolution}")
    ctrl.reset_camera

# Initialize application layout/ui
with SinglePageLayout(server) as layout:
    # Content setup
    with layout.content:
        with vuetify.VContainer(fluid=True, classes="pa-0 fill-height"):
            with vtk.VtkView() as vtk_view:               # vtk.js view for local rendering
                ctrl.reset_camera = vtk_view.reset_camera # Bind method to controller
                with vtk.VtkGeometryRepresentation():     # Add representation to vtk.js view
                    vtk.VtkAlgorithm(                     # Add ConeSource to representation
                        vtkClass="vtkConeSource",           # Set attribute value with no JS eval
                        state=("{ resolution }",)           # Set attribute value with JS eval
                    )


def show(server): 
    jupyter.show(server)

    return server.state
```
cell 2: 
```python
viewer = show(server)
```
cell 3: 
```python
viewer.resolution = 4
```
(The goal is to edit cell 3 such that the application running from cell 2 is automatically updated)

Thanks for the help!

**Comment by jourdain** [comment, marked as answer]
Here is a better way of doing apps for Jupyter

## Disable trame v3 warning

```python
import os

os.environ["TRAME_DISABLE_V3_WARNING"] = "1"
```

## Create app as class

```python

from trame.app import get_server
from trame.ui.vuetify import SinglePageLayout  
from trame.widgets import vuetify, vtk         
from trame.decorators import TrameApp, change


@TrameApp()
class Cone:
    def __init__(self, name=None):
        self.server = get_server(name)
        self.server.client_type = "vue2"
        self._ui = None

        # Build UI
        self.ui

    @property    
    def state(self):
        return self.server.state

    @property    
    def ctrl(self):
        return self.server.controller

    @property
    def resolution(self):
        return self.state.resolution

    @resolution.setter
    def resolution(self, v):
        with self.state:
            self.state.resolution = v

    def reset_resolution(self):
        self.resolution = 6

    @change("resolution")
    def _resolution_change(self, resolution, **kwargs):
        print("New resolution", resolution)

    @property
    def ui(self):
        if self._ui is None:
            with SinglePageLayout(self.server) as layout:
                self._ui = layout
                with layout.toolbar:
                    vuetify.VSpacer()
                    vuetify.VSlider(v_model=("resolution", 6), min=3, max=60, hide_details=True)
                    vuetify.VBtn("Reset", click=self.reset_resolution)
                with layout.content:
                    with vuetify.VContainer(fluid=True, classes="pa-0 fill-height"):
                        with vtk.VtkView() as vtk_view:
                            self.ctrl.reset_camera = vtk_view.reset_camera 
                            with vtk.VtkGeometryRepresentation():     
                                vtk.VtkAlgorithm(                     
                                    vtkClass="vtkConeSource",           
                                    state=("{ resolution }",)           
                                )

        return self._ui
```

## Jupyter usage and initialization

```python
cone = Cone()
await cone.ui.ready
cone.ui
```

## Changing state within Jupyter cell

```python
cone.resolution = 4
```

<img width="1552" alt="Screenshot 2023-08-23 at 8 57 52 AM" src="https://github.com/Kitware/trame/assets/294518/4c5731c6-ee26-4665-95e5-6e7e2e9340b3">

**Comment by cardinalgeo** [comment]
Seb this is awesome, thanks! Let me see if I'm getting this right: is the big enabler the use of the context manager defined in the resolution setter? 

And on another note, how does this new pattern fit in with the `jupyter.show` approach previously used (i.e., as defined [here](https://github.com/Kitware/trame/blob/861b60718798cca2db292e65e6ad39106ba75ccd/trame/app/jupyter.py#L12))? Is this meant to replace that?

And finally, why have you used the `@TrameApp` decorator in this case? I see from the documentation that it "...automatically register[s] methods to state.change, controller, trigger and life_cycle", but I'm not quite seeing which methods *need* to be registered here (on first pass, removing the decorator doesn't seem to change how the app behaves in the jupyter notebook). Why is it needed/beneficial here? Or is it just new best practice? 

Per usual, thank you for your quick, helpful response.

**Reply by jourdain** [reply]
So yes, when you are outside the normal execution in trame, using the state context manager trigger a flush on exit of any modified state value. This is required during async call or here in the jupyter context. You can also just call flush() on the state, but I feel the context manager is more elegant. 

So the new approach is more elegant and actually more correct from the point of view of Jupyter. In truth, they are doing the same thing except that we've extended the trame layout to be natively supported by Jupyter ([just by adding a few methods](https://github.com/Kitware/trame-client/blob/master/trame_client/ui/core.py#L143-L171)). The key part is the split with the await section. 

For that use case, you don't need state change listener so it is really not needed. But I wanted to illustrate how you can fully access trame capabilities within a class structure and decorators. So just a new best practice and skip calls like `state.change('something')(self.on_something_changed)` in a class constructor.

**Comment by cardinalgeo** [comment]
Perfect, this helps quite a bit! One last question: I'd like to be able to open the app either inline or in the browser. I've tried to set it up by extending your example with the following option: 

```python
        if self.inline == False:
            self.server.start(
                exec_mode="task",
                port=8080,
                open_browser=True, 
                show_connection_info=False,
                disable_logging=True,
                timeout=0,
                )
```

And while this works if I then go to the specified port, I'd like it to instead automatically open up the browser as is done when I run this from the command line. Am I missing something here that would enable that?

**Comment by jourdain** [comment]
The `exec_mode="task"` disable the [open browser](https://github.com/Kitware/trame-server/blob/master/trame_server/core.py#L543). But you could redo [that](https://github.com/Kitware/trame-server/blob/master/trame_server/core.py#L548) on your side.

**Reply by cardinalgeo** [reply]
That did the trick — I just added `self.ctrl.on_server_ready.add(lambda **kwargs: open_browser(self.server))` after `self.server.start(...`. Thanks Seb :)
